File: third-party/thrift/src/thrift/lib/go/thrift/serializer_test.go
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package thrift

import (
	"crypto/rand"
	"fmt"
	"reflect"
	"testing"

	"github.com/facebook/fbthrift/thrift/lib/go/thrift/dummy"
)

func TestGiantStructSerialization(t *testing.T) {
	// Serializer should be able to serialize a struct
	// much larger than the default size memory buffer.

	ser := NewBinarySerializer()

	giantByteBlob := make([]byte, defaultBufferSize*10)
	_, err := rand.Read(giantByteBlob)
	if err != nil {
		t.Fatalf("failed to rand read: %v", err)
	}
	val := dummy.NewDummyStruct1().
		SetField8(giantByteBlob).
		SetField9("giant_struct")
	pay, err := ser.Write(val)
	if err != nil {
		t.Fatalf("failed to write: %v", err)
	}

	var res dummy.DummyStruct1
	err = DecodeBinary(pay, &res)
	if err != nil {
		t.Fatalf("failed to decode: %v", err)
	}
	if !reflect.DeepEqual(*val, res) {
		t.Fatalf("values are not equal: %+v != %+v", *val, res)
	}
}

func TestConsequentSerialization(t *testing.T) {
	// A single Serializer instance should be able to
	// perform multiple sequential serializations.

	ser := NewBinarySerializer()

	for i := range 1000 {
		val := dummy.NewDummyStruct1().
			SetField9(fmt.Sprintf("sequential_test_value_%d", i))
		pay, err := ser.Write(val)
		if err != nil {
			t.Fatalf("failed to write: %v", err)
		}

		var res dummy.DummyStruct1
		err = DecodeBinary(pay, &res)
		if err != nil {
			t.Fatalf("failed to decode: %v", err)
		}
		if !reflect.DeepEqual(*val, res) {
			t.Fatalf("values are not equal: %+v != %+v", *val, res)
		}
	}
}

func TestSerializationBufferOwnership(t *testing.T) {
	// Serialized payload should reside in its own byte slice,
	// and not in the internal Serializer memory buffer.

	ser := NewBinarySerializer()
	prevRes := ([]byte)(nil)

	for range 10 {
		val := dummy.NewDummyStruct1()
		pay, err := ser.Write(val)
		if err != nil {
			t.Fatalf("failed to write: %v", err)
		}
		if reflect.ValueOf(pay).Pointer() == reflect.ValueOf(prevRes).Pointer() {
			t.Fatalf("payload pointers match")
		}
		prevRes = pay
	}
}

func BenchmarkBinarySerializer(b *testing.B) {
	// Run: buck run @mode/opt thrift/lib/go/thrift:thrift_test -- --test.bench=BenchmarkBinarySerializer -test.benchtime=100x

	ser := NewBinarySerializer()

	val := dummy.NewDummyStruct1().
		SetField1(int8(12)).
		SetField2(true).
		SetField3(int16(12345)).
		SetField4(int32(123456)).
		SetField5(int64(1234567)).
		SetField6(float32(1234.567)).
		SetField7(float64(123456.7)).
		SetField8([]byte{1, 2, 3, 4}).
		SetField9("hello")

	for b.Loop() {
		_, err := ser.Write(val)
		if err != nil {
			b.Fatalf("failed to write: %v", err)
		}
	}
}

File: third-party/thrift/src/thrift/lib/cpp2/transport/rocket/framing/test/SerializerTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <thrift/lib/cpp2/transport/rocket/framing/Serializer.h>

#include <gtest/gtest.h>

using namespace apache::thrift::rocket;

class HeaderSerializerTest : public testing::Test {
 protected:
  static constexpr size_t kMaxBufferSize = 64;
  uint8_t buffer[kMaxBufferSize];
  HeaderSerializer serializer{buffer, kMaxBufferSize};
};

TEST_F(HeaderSerializerTest, WriteBE_Success) {
  uint32_t value = 0x12345678;
  size_t bytesWritten = serializer.writeBE(value);

  EXPECT_EQ(bytesWritten, sizeof(value));
  EXPECT_EQ(serializer.result().size(), sizeof(value));
  EXPECT_EQ(
      *reinterpret_cast<const uint32_t*>(serializer.result().data()),
      folly::Endian::big(value));
}

TEST_F(HeaderSerializerTest, WriteBE_BufferFull) {
  uint64_t value = 0x123456789ABCDEF0;
  // Fill the buffer with some data to make it almost full
  size_t offset = kMaxBufferSize / sizeof(size_t);
  for (size_t i = 0; i < offset; ++i) {
    size_t bytesWritten = serializer.writeBE(i);
    EXPECT_EQ(bytesWritten, sizeof(size_t));
  }

  size_t bytesWritten = serializer.writeBE(value);

  // The buffer is expected to be full, so no bytes should be written
  EXPECT_EQ(bytesWritten, 0);
}

TEST_F(HeaderSerializerTest, WriteFrameOrMetadataSize_Success) {
  size_t nbytes = 0x12345;
  size_t bytesWritten = serializer.writeFrameOrMetadataSize(nbytes);

  EXPECT_EQ(bytesWritten, HeaderSerializer::kBytesForFrameOrMetadataLength);
  EXPECT_EQ(
      serializer.result().size(),
      HeaderSerializer::kBytesForFrameOrMetadataLength);
  EXPECT_EQ(serializer.result().data()[0], (nbytes >> 16) & 0xFF);
  EXPECT_EQ(serializer.result().data()[1], (nbytes >> 8) & 0xFF);
  EXPECT_EQ(serializer.result().data()[2], nbytes & 0xFF);
}

TEST_F(HeaderSerializerTest, WriteFrameOrMetadataSize_BufferFull) {
  // Fill the buffer with some data to make it almost full
  size_t offset = kMaxBufferSize / sizeof(size_t);
  for (size_t i = 0; i < offset; ++i) {
    size_t bytesWritten = serializer.writeBE(i);
    EXPECT_EQ(bytesWritten, sizeof(size_t));
  }

  size_t nbytes = 0x12345;
  size_t bytesWritten = serializer.writeFrameOrMetadataSize(nbytes);
  EXPECT_EQ(bytesWritten, 0);
}

TEST_F(HeaderSerializerTest, WriteFrameTypeAndFlags_Success) {
  FrameType frameType = FrameType::REQUEST_RESPONSE;
  Flags flags;
  flags.complete();

  size_t bytesWritten = serializer.writeFrameTypeAndFlags(frameType, flags);

  EXPECT_EQ(bytesWritten, sizeof(uint16_t));
  EXPECT_EQ(serializer.result().size(), sizeof(uint16_t));
  uint16_t resultValue =
      *reinterpret_cast<const uint16_t*>(serializer.result().data());
  uint16_t expectedValue = (static_cast<uint16_t>(frameType) << Flags::kBits) |
      static_cast<uint16_t>(flags);
  EXPECT_EQ(resultValue, folly::Endian::big(expectedValue));
}
